cluster/cluster.go

```diff
@@ -32,15 +32,58 @@ type Peer struct {
 	stopc  chan struct{}
 	readyc chan struct{}
 
+	peerLock    sync.RWMutex
+	peers       map[string]peer
+	failedPeers []peer
+
+	failedReconnectionsCounter     prometheus.Counter
+	successfulReconnectionsCounter prometheus.Counter
+	peerLeaveCounter               prometheus.Counter
+	peerUpdateCounter              prometheus.Counter
+	peerJoinCounter                prometheus.Counter
+
 	logger log.Logger
 }
 
+// peer is an internal type used for bookkeeping. It holds the state of peers
+// in the cluster.
+type peer struct {
+	status    PeerStatus
+	leaveTime time.Time
+
+	*memberlist.Node
+}
+
+// PeerStatus is the state that a peer is in.
+type PeerStatus int
+
+const (
+	StatusNone PeerStatus = iota
+	StatusAlive
+	StatusFailed
+)
+
+func (s PeerStatus) String() string {
+	switch s {
+	case StatusNone:
+		return "none"
+	case StatusAlive:
+		return "alive"
+	case StatusFailed:
+		return "failed"
+	default:
+		panic(fmt.Sprintf("unknown PeerStatus: %d", s))
+	}
+}
+
 const (
-	DefaultPushPullInterval = 60 * time.Second
-	DefaultGossipInterval   = 200 * time.Millisecond
-	DefaultTcpTimeout       = 10 * time.Second
-	DefaultProbeTimeout     = 500 * time.Millisecond
-	DefaultProbeInterval    = 1 * time.Second
+	DefaultPushPullInterval  = 60 * time.Second
+	DefaultGossipInterval    = 200 * time.Millisecond
+	DefaultTcpTimeout        = 10 * time.Second
+	DefaultProbeTimeout      = 500 * time.Millisecond
+	DefaultProbeInterval     = 1 * time.Second
+	DefaultReconnectInterval = 10 * time.Second
+	DefaultReconnectTimeout  = 6 * time.Hour
```

Comment:
what should these values be?

Reply:
Reconnection should probably be indefinite, for as long as SD returns the AM.

Reply:
Peers only exist from starting the binary (one of the `--cluster.peer` flags), or as an instance that connects to a running AM later. AM doesn't do any form of SD lookup to find its peers, so I think there needs to be some form of timeout since we have no way of knowing if a former peer is unreachable or has ceased to exist.